### DISCUSSION TITLE
Access Java enum with same name as static method

### DISCUSSION THREAD

**Discussion by mrckzgl** [jruby/jruby#8187] [Q&A] [answered]
Hey, we want to access the enums in this class: https://tinkerpop.apache.org/javadocs/current/core/org/apache/tinkerpop/gremlin/structure/VertexProperty.Cardinality.html from within JRuby.
Before an upgrade of the gremlin jar this worked with: `org.apache.tinkerpop.gremlin.structure.VertexProperty::Cardinality::single`, but after the upgrade, static methods with the same name as the enums were added to the class (e.g. https://tinkerpop.apache.org/javadocs/current/core/org/apache/tinkerpop/gremlin/structure/VertexProperty.Cardinality.html#single(java.lang.Object). Now if we try to access the enums, JRuby complains about a missing parameter, which indicates that it instead accesses the static method. 

How to distinguish between static enum and static method?

**Comment by headius** [comment]
> but after the upgrade

Just to be clear, this happened when you upgraded the gremlin jar, yes?

This is a tricky one; given that they have methods and enum values with the same way, I'm not sure how we can represent both at the same time in JRuby. I'll try to loop in some other JRuby folks to discuss.

**Reply by mrckzgl** [reply]
Jep exactly, upgrade of gremlin jar. No upgrade of JRuby.

**Comment by enebo** [comment]
Minimized example:

```java
public enum A {
    one,
    two,
    three,
    Four;

    public static void Four(int i) {}
    public static void one(int i) {}
}
```


```ruby
CLASSPATH=. jruby -I. -e 'p Java::A::const_get(:one)'  # boom (lowercase...boooooo)
```

but

```ruby
CLASSPATH=. jruby -I. -e 'p Java::A::const_get(:Four)'  # yay
```

So I think for JavaClasses we relax const_get to allow non-capitalized forms would fix this but also require a little metaprogramming.  I think this is ok in this case as it is atypical to use non-const enum values AND happen to have a static method with the same name.

**Comment by headius** [comment, marked as answer]
I've filed a bug for this in https://github.com/jruby/jruby/issues/8188.

A workaround for you for now would be to use reflection to access the enum values:

```
[] jruby $ cat Blah.java
public enum Blah {
    blah1, blah2;

    public static void blah1() {}
}
[] jruby $ jruby -e 'p Java::Blah.blah1, Java::Blah.blah2'
nil
blah2
[] jruby $ jruby -e 'p Java::Blah.java_class.get_field("blah1").get(nil), Java::Blah.blah2'
blah1
blah2
```

You should be able to take this further and define Ruby capitalized constants for those values, and just use the constants in your code.

**Reply by mrckzgl** [reply]
Thanks alot, that works nicely.
